pkg/client/fake/client_test.go

```diff
@@ -1877,13 +1873,136 @@ var _ = Describe("Fake client", func() {
+		cl := NewClientBuilder().WithObjects(&corev1.Pod{ObjectMeta: metav1.ObjectMeta{
+			Namespace: "default",
+			Name:      "foo",
+		}}).Build()
+
+		var pod corev1.Pod
+		Expect(cl.Get(context.Background(), client.ObjectKey{Namespace: "default", Name: "foo"}, &pod)).NotTo(HaveOccurred())
+
+		Expect(pod.TypeMeta).To(Equal(metav1.TypeMeta{}))
+	})
+
+	It("should leave typemeta empty on typed list", func() {
+		cl := NewClientBuilder().WithObjects(&corev1.Pod{ObjectMeta: metav1.ObjectMeta{
+			Namespace: "default",
+			Name:      "foo",
+		}}).Build()
+
+		var podList corev1.PodList
+		Expect(cl.List(context.Background(), &podList)).NotTo(HaveOccurred())
+		Expect(podList.ListMeta).To(Equal(metav1.ListMeta{}))
+		Expect(podList.Items[0].TypeMeta).To(Equal(metav1.TypeMeta{}))
+	})
+
+	It("should be able to Get an object that has pointer fields for metadata", func() {
+		schemeBuilder := &scheme.Builder{GroupVersion: schema.GroupVersion{Group: "test", Version: "v1"}}
+		schemeBuilder.Register(&WithPointerMeta{}, &WithPointerMetaList{})
+		scheme := runtime.NewScheme()
+		Expect(schemeBuilder.AddToScheme(scheme)).NotTo(HaveOccurred())
+
+		cl := NewClientBuilder().
+			WithScheme(scheme).
+			WithObjects(&WithPointerMeta{ObjectMeta: &metav1.ObjectMeta{
+				Name: "foo",
+			}}).
+			Build()
+
+		var object WithPointerMeta
+		Expect(cl.Get(context.Background(), client.ObjectKey{Name: "foo"}, &object)).NotTo(HaveOccurred())
+	})
+
+	It("should be able to List an object type that has pointer fields for metadata", func() {
+		schemeBuilder := &scheme.Builder{GroupVersion: schema.GroupVersion{Group: "test", Version: "v1"}}
+		schemeBuilder.Register(&WithPointerMeta{}, &WithPointerMetaList{})
+		scheme := runtime.NewScheme()
+		Expect(schemeBuilder.AddToScheme(scheme)).NotTo(HaveOccurred())
+
+		cl := NewClientBuilder().
+			WithScheme(scheme).
+			WithObjects(&WithPointerMeta{ObjectMeta: &metav1.ObjectMeta{
+				Name: "foo",
+			}}).
+			Build()
+
+		var objectList WithPointerMetaList
+		Expect(cl.List(context.Background(), &objectList)).NotTo(HaveOccurred())
+		Expect(objectList.Items).To(HaveLen(1))
+	})
+
+	It("should be able to List an object type that has pointer fields for metadata with no results", func() {
+		schemeBuilder := &scheme.Builder{GroupVersion: schema.GroupVersion{Group: "test", Version: "v1"}}
+		schemeBuilder.Register(&WithPointerMeta{}, &WithPointerMetaList{})
+		scheme := runtime.NewScheme()
+		Expect(schemeBuilder.AddToScheme(scheme)).NotTo(HaveOccurred())
+
+		cl := NewClientBuilder().
+			WithScheme(scheme).
+			Build()
+
+		var objectList WithPointerMetaList
+		Expect(cl.List(context.Background(), &objectList)).NotTo(HaveOccurred())
+		Expect(objectList.Items).To(BeEmpty())
+	})
 })
 
+type WithPointerMetaList struct {
+	*metav1.ListMeta
+	*metav1.TypeMeta
+	Items []*WithPointerMeta
+}
+
+func (t *WithPointerMetaList) DeepCopy() *WithPointerMetaList {
+	l := &WithPointerMetaList{
+		ListMeta: t.ListMeta.DeepCopy(),
```

Comment:
nit: does it make sense to also cover the nil cases for ListMeta (and ObjectMeta below)

Reply:
The `DeepCopy` implementations do nothing when they copy from a nil object which means that we implicitly have that for the `ListMeta` case. For the `ObjectMeta` case that isn't possible, because you can not access an object without specyfing at least a name anyways.

Reply:
Makes sense, thx! Didn't really think about that the Meta objects could be just nil safe :)